### PR TITLE
Implementation/graph summarization for hashtag recommendation

### PR DESCRIPTION
**Pull request by kacper1095**

### CODE REVIEW

File: tweet_recommendations/data_processing/tagged_files_to_dataframe.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import dask.bag as db
+import pandas as pd
+from dask.diagnostics import ProgressBar
+
+parser = argparse.ArgumentParser(
+    description="Converts tagged files by tagger toyger to dataframe with an id and lemmas")
+parser.add_argument("--dir", help="Str path to zip file")
+parser.add_argument("--out", help="Output file name with `p` extension")
+
+args = parser.parse_args()
```
Comment (mkosturek) [inline tweet_recommendations/data_processing/tagged_files_to_dataframe.py:16]: Może lepiej żeby to było w `main`, chociaż nie wiem czy to robi różnicę :)

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        similarity_rank_vertices = np.squeeze(similarity_rank_vertices.toarray())
+
+        best_indices = np.argsort(-similarities * similarity_rank_vertices)
+        result = self._hashtag_labels[best_indices].tolist()
+        return self.post_process_result(result)
+
+    def post_process_result(self, result: List[str]):
+        """
+        Removes ending suffix that was used to distinguish between nodes with the same name but different category.
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:242]: ❤️ Bardzo szanuję za dokumentację nawet w tak drobnych metodach...

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        """
+        if isinstance(x, str):
+            x = get_wcrft2_results_for_text(x)
+        if isinstance(x, list):
+            x = ' '.join(x)
+
+        # as in fit, vectorizer has normalization inside ...
+        tf_idf_vector = self._hashtags_tf_idf_vectorizer.transform([x])
+
+        # ... so this simplifies to cosine similarity
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py]: zmieniłbym ten komentarz `... so this simplifies to cosine similarity` na `... so this simplifies calculation of cosine similarity - no normalisation required`
bo trochę wprowadza w błąd, sugerując że z jakiegoś innego wzorku upraszcza nam się coś do cosine similarity, chociaż to nie o to chodzi

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        nb_iteration = 0
+
+        while True:
+            if self.verbose:
+                print("Step: {}".format(nb_iteration + 1))
+
+            similarity_rank_vertices = self.damping_factor * self._transition_matrix.dot(similarity_rank_vertices) + (
+                    1 - self.damping_factor) * preference_vector
+
+            diff = np.abs(np.sum(similarity_rank_vertices - previous_similarity_rank_vertices))
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py]: hmm.. nie wiem czy nie lepiej byłoby zrobić na odwrót. Zamiast `np.abs(np.sum(...))`, zrobić: `np.sum(np.abs(...))`
W tej chwili może dojść do sytuacji:
`previous_similarity_rank_vertices = [1,-1]`
`similarity_rank_vertices = [-1,1]` ( zmiana o 180 stopni)
`d = (similarity_rank_vertices - previous_similarity_rank_vertices) = [-2, 2]`
`s = sum(d) = 0`
`diff = abs(s) = 0`

w tym co proponuję wyszłoby:
`a = abs(d) = [2, 2]`
`diff = sum(a) = 4`
Czyli wykażemy, że jeszcze się coś zmienia i jeszcze nie zbiegło

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            if nb_iteration > 0 and diff < self.minimal_random_walk_change_difference_value:
+                if self.verbose:
+                    print("Converged with error: {:.6f}".format(diff))
+                break
+
+            previous_similarity_rank_vertices = similarity_rank_vertices
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py]: Dla przejrzystości przypisanie `previous_similarity_rank_vertices = similarity_rank_vertices` robiłbym raczej na początku kroku pętli a nie na końcu, chyba że są jakieś ku temu powody, których nie widzę :)

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+            nb_iteration += 1
+
+            if nb_iteration > self.max_iterations:
+                if self.verbose:
+                    print("Random walk did not converged, current error: {:.6f}".format(diff))
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py]: `did not converged` -> `did not converge` :P

File: tweet_recommendations/other_methods/lemmatizer.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import os
+
+import pandas as pd
+from sklearn.base import TransformerMixin, BaseEstimator
+
+
+class TaggerToygerLemmatizer(TransformerMixin, BaseEstimator):
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/lemmatizer.py:7]: Rozumiem że to kod, który już wcześniej był gdzieś wykorzystywany, ale zaginął i raczej nie ma potrzeby go przeglądać? :P

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        :param verbose: bool. Print progressbar while iterating through nodes.
+        :return: None.
+        """
+        new_graph = self.graph.copy()
+        for node in tqdm.tqdm(self.graph.nodes(), disable=not verbose):
+            if self.graph.node[node]["type"] == "hashtag":
+                for neighbour in self.graph.neighbors(node):
+                    if self.graph.node[neighbour]["type"] == "username":
+                        for other_node in self.graph.neighbors(neighbour):
+                            if self.graph.node[other_node]["type"] == "hashtag" and not self.graph.has_edge(node,
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py]: nie ma sprawdzania czy `other node == node` czyli czy nie łączymy hashtaga samego ze sobą.
Wiem, że potem w incidence matrix diagonala jest zerowana, ale może warto już tutaj o to zadbać żeby mieć spójny graf i incidence matrix. - Propozycja, nie przymus.

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+        self._refine_matrix_with_additional_connections(self.verbose)
+
+        self._hashtag_labels = np.asarray(list(self._hashtag_labels))
+        self._users_labels = np.asarray(list(self._users_labels))
+        self._tweet_labels = np.asarray(list(self._tweet_labels))
+
+        if self.verbose:
+            print("Building incidence matrix ...")
+        incidence_matrix = self._get_binary_incidence_matrix()[:len(self._hashtag_labels), len(self._hashtag_labels):]
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:156]: `[:len(self._hashtag_labels), len(self._hashtag_labels):]` - zastanawia mnie to indeksowanie, że na jednym wymiarze jest `: len` a na drugim `len :` - czy to na pewno poprawne?

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+        weighted_adjacency_matrix_of_tags = incidence_matrix.dot(incidence_matrix.T)
+        weighted_adjacency_matrix_of_tags.setdiag(0)
+
+        if self.verbose:
+            print("Building hashtag graph ...")
+
+        hashtag_graph = nx.from_scipy_sparse_matrix(weighted_adjacency_matrix_of_tags)
+
+        weighted_degree = np.asarray(list(map(itemgetter(1), hashtag_graph.degree(weight="weight"))))
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:165]: rozumiem że `nx.from_scipy_sparse_matrix(weighted_adjacency_matrix_of_tags)` buduje graf gdzie każda krawędź m atrybut "weight" zgodny z wartościami w `weighted_adjacency_matrix_of_tags`? i stąd bierze się `hashtag_graph.degree(weight="weight")` ?

I `hashtag_graph.degree(...)` zwraca wyniki w ustalonej kolejności węzłów, a nie przetasowuje je?

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+        preference_vector = np.zeros((len(self._hashtag_labels, )))[..., np.newaxis]
+        preference_vector[query_hashtag_index] = 1
+        preference_vector = sps.csr_matrix(preference_vector, shape=preference_vector.shape, dtype=np.float32)
+
+        similarity_rank_vertices = preference_vector
+        previous_similarity_rank_vertices: np.ndarray = preference_vector
+        nb_iteration = 0
+
+        while True:
```
Comment (mkosturek) [inline tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:218]: Możnaby wydzielić random walka do osobnej metody, ale jeśli wyjdzie dużo argumentów i dużo wyjść to może nie ma to sensu... wtedy ew komentarz że tu się zaczyna random walk

**Review by mkosturek — CHANGES_REQUESTED** [review]
Generalnie: 👍 🥇 
Kilka uwag estetyczno-clean coderskich
Chyba tylko jedna uwaga co do działania - przy sprawdzaniu warunku zbiegania random walka, może być istotna - dlatego "Request Changes".

File: tweet_recommendations/other_methods/lemmatizer.py (hunk 7 above)
Comment (kacper1095) [reply tweet_recommendations/other_methods/lemmatizer.py:7]: Bardziej to służy do wykorzystania lemmatów zrobiony przez TaggeraToygera niż do działania taggera toygera

File: tweet_recommendations/data_processing/tagged_files_to_dataframe.py (hunk 1 above)
Comment (kacper1095) [reply tweet_recommendations/data_processing/tagged_files_to_dataframe.py:16]: O tyle nie, że mógłbyś zaimportować ten skrypt i użyć jedną z metod tam zawartych. Argument Parser trochę Cię blokuje przed tym w tym momencie i przynajmniej u mnie oznacza tyle, że skrypt jest samowystarczalny i nie powinien być używany nigdzie indziej

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py (hunk 2 above)
Comment (kacper1095) [reply tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:242]: ^_^

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py (hunk 6 above)
Comment (kacper1095) [reply tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py]: Pytanie: w emfazie też DID niweluje czas przeszły czasownika?

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py (hunk 9 above)
Comment (kacper1095) [reply tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:156]: Poprawnie, bo chcesz uzyskać macierz: 
image
Jakby nie patrzeć, to w kodzie `_get_binary_incidence_matrix` macierz jest tworzona według labelek na podstawie `nodelist`:
image
Więc żeby to zadziałało, to z wierszy trzeba wziąć pierwsze H elementów a z kolumn T + U

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py (hunk 10 above)
Comment (kacper1095) [reply tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:165]: Na podstawie przykładu: https://networkx.github.io/documentation/stable/reference/classes/generated/networkx.Graph.degree.html sądzę, że nie. Sprawdzałem osobiście, zawsze listuje je od 0 do N.

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py (hunk 4 above)
Comment (kacper1095) [reply tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py]: Racja, zawsze piszę jak mówisz, a tutaj wyszło pisanie na szybko :)

File: tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py (hunk 9 above)
Comment (mkosturek) [reply tweet_recommendations/other_methods/graph_summarization_for_hashtag_recommendation.py:156]: Już rozumiem, zgadza się :+1:

**Review by mkosturek — APPROVED** [review]
Wszystko gra

File: tweet_recommendations/validators/graph_summarization_validator.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+lemmatizer = TaggerToygerLemmatizer("data/processed/lemmas.p")
+the_graph = GraphSummarizationMethod(max_iterations=100, minimal_random_walk_change_difference_value=1e-4,
+                                     verbose=True, damping_factor=0.8)
+
+with open("data/source_data/original_tweets.p", "rb") as f:
+    original_tweets = pkl.load(f)
+
+tweets_with_lemmas = lemmatizer.fit_transform(original_tweets, verbose=True)
+the_graph.fit(tweets_with_lemmas, minimal_hashtag_occurence=3)
+print(the_graph.transform("Wiozę szwagra na wybory")[:10])
```
Comment (jmichalowski12) [inline tweet_recommendations/validators/graph_summarization_validator.py:15]: Fajnie, że są te printy, ale nie wiem, czy powinny trafić do mastera)

**Comment by jmichalowski12** [comment]
Co do uwag estetycznych to proponuję sobie [black'a](https://github.com/ambv/black) zainstalować.
On rozwiązuje wątpliwości i ręczne pilnowanie zgodności z PEP8 - nie uważam, że przeformatowany kod trzymający się pepa 1 do 1 jest najczytelniejszy - ale rozwiąże nam to problem spójności formatowania plików

Inna uwaga jest to o czym rozmawialiśmy - to kodzenie w stylu sklearnowym dla mnie zaciemnia trochę działanie całości, ale biorąc pod uwagę, że ktoś może zechcieć z tego korzystać to dobrze jest taką strukturę utrzymać, żeby nikt się nie musiał zastanawiać jak z tego korzystać.

**Comment by kacper1095** [comment]
Sam pilnuję formatowania PEP8, w pycharmie jest taka opcja. Na razie na podstawie widzianego kodu to odnoszę wrażenie, że black to najgorszy sort formatowania i nie będę tego instalował. W dodatku on nie przestrzega PEP8. Co najwyżej mogę się zgodzić na flake8, który już tego formatowania przestrzega

**Review by jmichalowski12 — APPROVED** [review]
